src/server/page_server.cpp

```diff
@@ -237,7 +206,128 @@ void page_server::on_data_page_read_result (const FILEIO_PAGE *io_page, int erro
 	}
     }
 
-  m_active_tran_server_conn->push (page_to_tran_request::SEND_DATA_PAGE, std::move (message));
+  m_conn->push (page_to_tran_request::SEND_DATA_PAGE, std::move (message));
+}
+
+void page_server::set_active_tran_server_connection (cubcomm::channel &&chn)
+{
+  assert (is_page_server ());
+
+  chn.set_channel_name ("ATS_PS_comm");
+  er_log_debug (ARG_FILE_LINE, "Active transaction server connected to this page server. Channel id: %s.\n",
+		chn.get_channel_id ().c_str ());
+
+  if (m_active_tran_server_conn != nullptr)
+    {
+      // ATS must have crashed without disconnecting from us. Destroy old connection to create a new one.
+      disconnect_tran_server (m_active_tran_server_conn.get ());
+    }
+  assert (m_active_tran_server_conn == nullptr);
+  m_active_tran_server_conn.reset ( new connection_handler (chn, this));
+}
+
+void page_server::set_passive_tran_server_connection (cubcomm::channel &&chn)
+{
+  assert (is_page_server ());
+
+  chn.set_channel_name ("PTS_PS_comm");
+  er_log_debug (ARG_FILE_LINE, "Passive transaction server connected to this page server. Channel id: %s.\n",
+		chn.get_channel_id ().c_str ());
+
+  m_passive_tran_server_conn.emplace_back (new connection_handler (chn, this));
+}
+
+void page_server::disconnect_tran_server (void *conn)
+{
+  if (conn == nullptr)
+    {
+      if (m_active_tran_server_conn == nullptr)
+	{
+	  er_log_debug (ARG_FILE_LINE, "Page server was never connected with an active transaction server.\n");
+	  return;
+	}
+      if (m_passive_tran_server_conn.empty ())
+	{
+	  er_log_debug (ARG_FILE_LINE, "Page server was never connected with an passive transaction server.\n");
+	  return;
+	}
+    }
```

Comment:
conn should never be null.
```suggestion
  assert (conn != nullptr);
```